camel/types/model_type.py

```diff
@@ -0,0 +1,141 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+from typing import ClassVar, Dict, Union
+
+from camel.types import PredefinedModelType
+
+
+class ModelType:
+    _cache: ClassVar[Dict[str, "ModelType"]] = {}
+
+    def __new__(cls, *args, **kwargs) -> "ModelType":
+        if not args:
+            return super().__new__(cls)
+
+        # If args are passed in, meaning that this instance is being created
+        # by user with calling ModelType(...) directly
+        value: Union[PredefinedModelType, str] = args[0]
+        if isinstance(value, PredefinedModelType):
+            value_str = value.value
+        elif isinstance(value, str):
+            value_str = value
+        else:
+            raise ValueError(f"Invalid type for ModelType: {value}.")
+        if value_str in cls._cache:
+            return cls._cache[value_str]
+        instance = super().__new__(cls)
+        cls._cache[value_str] = instance
+        return instance
+
+    def __init__(self, value: Union[str, PredefinedModelType]) -> None:
+        # If type attr is set, then the instance is from cache
+        if hasattr(self, "type"):
+            return
+
+        if isinstance(value, PredefinedModelType):
+            self.type = value
+            self.value = value.value
+        elif isinstance(value, str):
+            try:
+                self.type = PredefinedModelType(value)
+            except ValueError:
+                self.type = PredefinedModelType.OPEN_SOURCE
```

Comment:
here if the model is not in our list then we set the type to opensource, actually it may not be the case, for example there are some commercial models supported by litellm

Reply:
Then are we going to add these models into our list also? I think in our previous discussion we think that we should actively maintain all the commercial models in the PredefinedModelTypes right

Reply:
Maybe name them as `CustomModel`?

Reply:
@WHALEEYE It's also hard to maintain all commercial models within our list, let's just change the naming here instead of using `OPEN_SOURCE `